File: server/EvalutionAlgoritmCpp/EvalutionAlgoritmCpp/EvulationAlgoritmCpp/Crossing.cpp
#include "Crossing.h"
#include "GetRndInteger.h"
#include <set>
#include <iostream>
using namespace std;
void Crossing(individ& schedule1, individ& schedule2, const vector<clas>& classes)
{
	
	int start = GetRndInteger(0, (classes.size() - 1));
	int end = GetRndInteger(start, (classes.size() - 1));
	for(int i = start;i<end;i++)
	{
		set<int>id_aud1;
		set<int> id_aud2;
		bool isFirst = false;
		for(assigned_group ag: classes[i].assigned_groups)
		{
			vector<schedule>& scheduleForGr1 = schedule1.scheduleForGroups.find(ag.id_group)->second;
			vector<schedule> tempT1,tempF1;
			for(schedule sch: scheduleForGr1)
			{
				if (sch.id_class != classes[i].id)
					tempT1.push_back(sch);
				else
					tempF1.push_back(sch);
			}
			scheduleForGr1 = tempT1;
			vector<schedule>& scheduleForGr2 = schedule2.scheduleForGroups.find(ag.id_group)->second;
			vector<schedule> tempT2, tempF2;
			for (schedule sch : scheduleForGr2)
			{
				if (sch.id_class != classes[i].id)
					tempT2.push_back(sch);
				else
					tempF2.push_back(sch);
			}
			scheduleForGr2 = tempT2;
			if(!tempF2.empty())
			scheduleForGr1.insert(scheduleForGr1.end(), tempF2.begin(), tempF2.end());
			if (!tempF1.empty())
			scheduleForGr2.insert(scheduleForGr2.end(), tempF1.begin(), tempF1.end());
			if(!isFirst)
			{
				isFirst = true;
				for(schedule sch:tempF1)
				{
					id_aud1.insert(sch.id_audience);
				}
				for (schedule sch : tempF2)
				{
					id_aud2.insert(sch.id_audience);
				}
			}
		}
		for(assigned_teacher at: classes[i].assigned_teachers)
		{
			
			vector<schedule>& scheduleForTeach1 = schedule1.scheduleForTeachers.find(at.id_teacher)->second;
			vector<schedule> tempT1, tempF1;
			for (schedule sch : scheduleForTeach1)
			{
				if (sch.id_class != classes[i].id)
					tempT1.push_back(sch);
				else
					tempF1.push_back(sch);
			}
			vector<schedule>& scheduleForTeach2 = schedule2.scheduleForTeachers.find(at.id_teacher)->second;
			vector<schedule> tempT2, tempF2;
			for (schedule sch : scheduleForTeach2)
			{
				if (sch.id_class != classes[i].id)
					tempT2.push_back(sch);
				else
					tempF2.push_back(sch);
			}
			scheduleForTeach1 = tempT1;
			scheduleForTeach2 = tempT2;
			if (!tempF2.empty())
				scheduleForTeach1.insert(scheduleForTeach1.end(), tempF2.begin(), tempF2.end());
			if (!tempF1.empty())
				scheduleForTeach2.insert(scheduleForTeach2.end(), tempF1.begin(), tempF1.end());
		}
		vector <schedule> temp1,temp2;
		for(int id_audience: id_aud1)
		{
			vector<schedule>& scheduleForAud1 = schedule1.scheduleForAudiences.find(id_audience)->second;
			vector<schedule> tempT1, tempF1;
			for(schedule sch:scheduleForAud1)
			{
				if (sch.id_class != classes[i].id)
					tempT1.push_back(sch);
				else
					tempF1.push_back(sch);
			}
			scheduleForAud1 = tempT1;
			if(!tempF1.empty())
			temp1.insert(temp1.end(), tempF1.begin(), tempF1.end());
		}
		for (int id_audience : id_aud2)
		{
			vector<schedule>& scheduleForAud2 = schedule2.scheduleForAudiences.find(id_audience)->second;
			vector<schedule> tempT2, tempF2;
			for (schedule sch : scheduleForAud2)
			{
				if (sch.id_class != classes[i].id)
					tempT2.push_back(sch);
				else
					tempF2.push_back(sch);
			}
			scheduleForAud2 = tempT2;
			if (!tempF2.empty())
			temp2.insert(temp2.end(), tempF2.begin(), tempF2.end());
		}
		
			for (int id_audience : id_aud1)
			{
				vector<schedule>temp;
				if(schedule2.scheduleForAudiences.find(id_audience)== schedule2.scheduleForAudiences.end())
				{
					vector<schedule> t;
					schedule2.scheduleForAudiences[id_audience] = t;
				}

				vector<schedule>& scheduleForAud2 = schedule2.scheduleForAudiences.find(id_audience)->second;
				for (schedule sch : temp1)
				{
					if (sch.id_audience == id_audience)
						temp.push_back(sch);
				}
				if (!temp.empty())
				scheduleForAud2.insert(scheduleForAud2.end(), temp.begin(), temp.end());
			}
			for (int id_audience : id_aud2)
			{
				vector<schedule>temp;
				if (schedule1.scheduleForAudiences.find(id_audience) == schedule1.scheduleForAudiences.end())
				{
					vector<schedule> t;
					schedule1.scheduleForAudiences[id_audience] = t;
				}
				vector<schedule>& scheduleForAud1 = schedule1.scheduleForAudiences.find(id_audience)->second;
				for (schedule sch : temp2)
				{
					if (sch.id_audience == id_audience)
						temp.push_back(sch);
				}
				if (!temp.empty())
				scheduleForAud1.insert(scheduleForAud1.end(), temp.begin(), temp.end());
			}
		}
	
}

File: server/Algorithms/EvalutionAlgorithm/Crossing.js
import GetRndInteger from "../Service/GetRndInteger.js";
import replacer from "./JSONReplacer.js";
import reviver from "./JSONReviver.js";

export default function Crossing(schedule1, schedule2, classes) {
  let s = GetRndInteger(0, classes.length - 1);
  let f = GetRndInteger(s, classes.length - 1);
  let population_child1 = JSON.parse(schedule1, reviver);
  let population_child2 = JSON.parse(schedule2, reviver);
  for (let i = s; i <= f; i++) {
    let id_audiences1 = new Set();
    let id_audiences2 = new Set();
    let isFirst = false;
    // Обмен занятий для групп
    classes[i].assigned_groups.forEach((ag) => {
      let tempScheduleForGroup1 = population_child1.scheduleForGroups.get(
        ag.id_group
      );
      let temp1 = [];
      tempScheduleForGroup1 = tempScheduleForGroup1.filter((sch) => {
        if (sch.id_class !== classes[i].id) return true;
        else {
          temp1.push(sch);
          return false;
        }
      });
      let tempScheduleForGroup2 = population_child2.scheduleForGroups.get(
        ag.id_group
      );
      let temp2 = [];
      tempScheduleForGroup2 = tempScheduleForGroup2.filter((sch) => {
        if (sch.id_class !== classes[i].id) return true;
        else {
          temp2.push(sch);
          return false;
        }
      });
      tempScheduleForGroup1.push(...temp2);
      tempScheduleForGroup2.push(...temp1);
      population_child1.scheduleForGroups.set(
        ag.id_group,
        tempScheduleForGroup1
      );
      population_child2.scheduleForGroups.set(
        ag.id_group,
        tempScheduleForGroup2
      );
      // Пройтись по занятиям что бы взять аудитории
      if (!isFirst) {
        isFirst = true;
        temp1.forEach((sc) => {
          id_audiences1.add(sc.id_audience);
        });
        temp2.forEach((sc) => {
          id_audiences2.add(sc.id_audience);
        });
      }
    });
    // Обмен занятий для учителей
    classes[i].assigned_teachers.forEach((at) => {
      let tempScheduleForTeacher1 = population_child1.scheduleForTeachers.get(
        at.id_teacher
      );
      let temp1 = [];
      tempScheduleForTeacher1 = tempScheduleForTeacher1.filter((sch) => {
        if (sch.id_class !== classes[i].id) return true;
        else {
          temp1.push(sch);
          return false;
        }
      });
      let tempScheduleForTeacher2 = population_child2.scheduleForTeachers.get(
        at.id_teacher
      );
      let temp2 = [];
      tempScheduleForTeacher2 = tempScheduleForTeacher2.filter((sch) => {
        if (sch.id_class !== classes[i].id) return true;
        else {
          temp2.push(sch);
          return false;
        }
      });
      tempScheduleForTeacher1.push(...temp2);
      tempScheduleForTeacher2.push(...temp1);
      population_child1.scheduleForTeachers.set(
        at.id_teacher,
        tempScheduleForTeacher1
      );
      population_child2.scheduleForTeachers.set(
        at.id_teacher,
        tempScheduleForTeacher2
      );
    });
    // Обмен занятий для аудитории
    let temp1 = [];
    id_audiences1.forEach((id_audience) => {
      let scheduleForAudience1 =
        population_child1.scheduleForAudiences.get(id_audience);
      let temp = [];
      scheduleForAudience1 = scheduleForAudience1.filter((sch) => {
        if (sch.id_class !== classes[i].id) return true;
        else {
          temp.push(sch);
          return false;
        }
      });
      population_child1.scheduleForAudiences.set(
        id_audience,
        scheduleForAudience1
      );
      temp1.push(...temp);
    });
    let temp2 = [];
    id_audiences2.forEach((id_audience) => {
      let scheduleForAudience2 =
        population_child2.scheduleForAudiences.get(id_audience);
      let temp = [];
      scheduleForAudience2 = scheduleForAudience2.filter((sch) => {
        if (sch.id_class !== classes[i].id) return true;
        else {
          temp.push(sch);
          return false;
        }
      });
      population_child2.scheduleForAudiences.set(
        id_audience,
        scheduleForAudience2
      );
      temp2.push(...temp);
    });
    id_audiences1.forEach((id_audience) => {
      let temp = temp1.filter((sch) => sch.id_audience === id_audience);
      let temp_sch = population_child2.scheduleForAudiences.get(id_audience);
      if (!temp_sch) temp_sch = [];
      temp_sch.push(...temp);
      population_child2.scheduleForAudiences.set(id_audience, temp_sch);
    });
    id_audiences2.forEach((id_audience) => {
      let temp = temp2.filter((sch) => sch.id_audience === id_audience);
      let temp_sch = population_child1.scheduleForAudiences.get(id_audience);
      if (!temp_sch) temp_sch = [];
      temp_sch.push(...temp);
      population_child1.scheduleForAudiences.set(id_audience, temp_sch);
    });
  }
  return {
    population_child1: JSON.stringify(population_child1, replacer),
    population_child2: JSON.stringify(population_child1, replacer),
  };
}
